File: numguess.fs
open System

// Random generator
let rnd = Random()
let random n = rnd.Next(n) + 1

// Reads a string from the console with default value
let read def =
    let s = Console.ReadLine()
    if s = "" then def else s

// Reads an int from the console
let readInt def =
    try
        (int (read ""), true)
    with
        | ex -> (def, false)

// Returns a message if guess is high or low
let highOrLow guess number =
    if fst guess < number then "Too low!"
    elif fst guess > number then "Too high!"
    else ""

// Returns custom message
let message tries limit =
    let max_tries = int (floor (log (float limit) / log 2.0)) + 1
    if tries = 1 then "You're one lucky bastard!"
    elif tries < max_tries then "You are the master of this game!"
    elif tries = max_tries then "You are a machine!"
    elif tries <= int (float max_tries * 1.1) then "Very good result!"
    elif tries <= limit then "Try harder, you can do better!"
    else "I find your lack of skill disturbing!"

// Main program
printf "Welcome to NumGuess F# version!\n\nEnter your name: "
let name = read "Player"

printf "\nWelcome %s, enter upper limit: " name
let limit = max (fst (readInt 10)) 10

let mutable again = true
while again do
    let mutable tries = 0
    let mutable guess = (0, true)
    let mutable number = random limit
    printfn "\nGuess my number between 1 and %d!\n" limit
    while number <> fst guess do
        printf "Guess: "
        guess <- readInt 0
        if not (snd guess) then
            printfn "That's just plain wrong."
        elif fst guess > limit || fst guess < 1 then
            printfn "Out of range."
        else
            tries <- tries + 1
            printfn "%s" (highOrLow guess number)
    printfn "Well done %s, you guessed my number from %d %s!" name tries (if tries = 1 then "try" else "tries")
    printfn "%s" (message tries limit)
    printf "Play again [y/N]? "
    again <- (read "y").ToLower() = "y"
printfn "\nOkay, bye."

File: numguess.hs
import System.IO
import System.Random

getName s = if length s == 0 then "Player" else s

getLimit :: String -> Int
getLimit [] = 10
getLimit s = max 10 (fst (readNumber s))

readNumber :: String -> (Int, Bool)
readNumber [] = (0, False)
readNumber s = if length r == 0 then readNumber [] else (fst (head r), True)
    where r = reads s :: [(Int, String)]

evaluate :: (Int, Bool) -> Int -> Int -> (String, Int, Bool)
evaluate (guess, valid) num limit
    | not valid              = ("That's just plain wrong.", 0, False)
    | num < 1 || num > limit = ("Out of range.", 0, False)
    | guess < num            = ("Too low!", 1, False)
    | guess > num            = ("Too high!", 1, False)
    | guess == num           = ("", 1, True)

customMessage :: Int -> Int -> String
customMessage tries limit
    | tries == 1          = "You're one lucky bastard!"
    | tries < maxTries    = "You are the master of this game!"
    | tries == maxTries   = "You are a machine!"
    | tries <= maxTries10 = "Very good result!"
    | tries <= limit      = "Try harder, you can do better!"
    | otherwise           = "I find your lack of skill disturbing!"
    where maxTries = floor (logBase 2 (fromIntegral limit)) + 1
          maxTries10 = floor (fromIntegral maxTries * 1.1)

playAgain :: String -> Bool
playAgain input
    | input == "Y" || input == "y" = True
    | otherwise                    = False

main = do
    hSetBuffering stdout NoBuffering
    putStr "Welcome to NumGuess Haskell version!\n\nEnter your name: "
    input <- getLine
    let name = getName input
    putStr ("\nWelcome " ++ name ++ ", enter upper limit: ")
    input <- getLine
    let limit = getLimit input
    play name limit

play name limit = do
    putStrLn ("\nGuess my number between 1 and " ++ show limit ++ "!\n")
    num <- randomRIO (1, limit)
    guess name limit num 0
    putStr "Play again [y/N]? "
    input <- getLine
    if playAgain input then
        play name limit
    else
        putStrLn "\nOkay, bye."

guess name limit num tries = do
    putStr "Guess: "
    input <- getLine
    let (message, add, finished) = evaluate (readNumber input) num limit
    putStrLn message
    if not finished then
        guess name limit num (tries + add)
    else do
        let t = tries + add
        putStr ("Well done " ++ name ++ ", you guessed my number from " ++ show t ++ " ")
        putStrLn (if t == 1 then "try!" else "tries!")
        putStrLn (customMessage t limit)
